### ISSUE TITLE
Frequently disconnected from Azure SQL database using AAD authentications

### ISSUE THREAD

**Issue by erapade** [microsoft/azuredatastudio#14136]
Issue Type: <b>Bug</b>

We are frequently disconnected from Azure SQL database using AAD authentication. Looking for the best way to track down the issue.

Working from home so no corp. FW should be involved and the FW settings on the SQL server are granting me access. But yes, we are also using a SaaS proxy-server from Zscaler, although, both the IP from Zscaler and my internet NAT IP are cleared in the Azure SQL server FW

Azure Data Studio version: azuredatastudio 1.25.2 (4d4917d328ab7a7526817a0f0ef03237a79aa54d, 2021-01-20T22:52:13.492Z)
OS version: Windows_NT x64 10.0.19042

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz (4 x 2904)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.84GB (3.52GB free)|
|Process Argv|--folder-uri file:///c%3A/Users/ds82395/source/repos/Sandvik%20GFIT%20BI%20-%20GIT%20-%20SQL%20Database|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (3)</summary>

Extension|Author (truncated)|Version
---|---|---
eltsnap-simple-data-flow|bit|1.0.1
profiler|Mic|0.11.0
poor-sql-formatter|WSR|0.1.0

</details>

**Comment by erapade** [comment]
This is the error message:
Login failed for user '<token-identified principal>'. Token is expired.

**Comment by kburtram** [comment]
@cssuh do we know how long the AAD tokens are expected to valid prior to expiration?  I assume this issue is just that we need to auto-renew the refresh token when expired?

**Comment by erapade** [comment]
Reading this: https://github.com/microsoft/azuredatastudio/issues/12165

I noticed my "azureTokenCache_azure_publicCloud" file was modified at 13:06 today and not it's 16:21 and I have lost access. I will delete that file and see if the issue goes away. En general, this happens for me maybe 10 times a day so it's very irritating.

**Comment by erapade** [comment]
I found this, don't know if this is related, but the change was happening approximate the same time as I started to get problems:
https://docs.microsoft.com/en-us/azure/active-directory/develop/configure-token-lifetimes#create-token-lifetime-policies-for-refresh-and-session-tokens
and this: https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-configurable-token-lifetimes#configurable-token-lifetime-properties-after-the-retirement

**Comment by erapade** [comment]
> "azureTokenCache_azure_publicCloud"

F.Y.I. Deleting the "azureTokenCache_azure_publicCloud" file didn't help, problem the problem reoccured after some time (1-2 hours)

**Comment by erapade** [comment]
@kburtram: Is there really now update on this. I guess all users working towards Azure SQL databases with AAD authentication will have this very same problem

**Comment by erapade** [comment]
@kburtram Just checking if this has been forgotten since it once was pointed to the Mars Release but wasn't included in that release

**Comment by kburtram** [comment]
@erapade the issue hasn't been forgotten.  We just have a lot of work in-progress so sometimes the issues take a longer to get to than originally intended.  @caohai is currently looking into the AAD refresh token expiration issue.

**Comment by erapade** [comment]
I also reported this using premier support, case number 2106230040004050. I have been told a fix shall be included in a August release but haven’t seen any change yet. Still hoping though.

**Comment by erapade** [comment]
I got the August version yesterday, but that did not include any fix for this issue

**Comment by caohai** [comment]
We've made some progress with AAD token expiration during the past few months. Query editor and notebook should not have this issue any more. They should be in the latest release, feel free to try it out and share if it fixes the issue for you.

There are still a few places this issue will occur and we are still working on them. You should be able to find the latest progress in this mega issue #17047.

**Comment by erapade** [comment]
The query editor now works better, thanks for this, and thanks for giving an update, very appreciated!!! 
When I today started to use the query windows I used yesterday, it did say disconnected, and I then needed to re-connect to the database. Sadly my old connection wasn't pre-filled-in, so I needed to choose the connections once again. I here see a risk that you connect to the wrong database and this could potentially cause serious problems

And sadly I could see that the main issue with connections in the object explorer now once again has been re-scheduled, this time from December to February. Considering I reported this in February 2021, and the problem started to be a real problem when Microsoft changed the default AAD token expiration them-self, this has been taken too long time to fix. 

Considering the query editor doesn't pre-fill the earlier connection data I guess that this is not the final solution on the problem, and thereby, this issue should not be closed if it was logged just as a query window problem. But this ticket was created to address the complete issue, not only the query window issues, and then I see no reason at all to close down this ticket if it's not just to create better statistics. Sorry for being harsh, but having had this issue for almost a year, I have the feeling of this not being prioritized as it should. My recommendation is to reopen this and not close it until my problem has been confirmed to be solved
